File: csharp/factory_method.cs
using System;

abstract class Animal
{
    public abstract string Speak();
}

class Dog : Animal
{
    public override string Speak()
    {
        return "Woof!";
    }
}

class Cat : Animal
{
    public override string Speak()
    {
        return "Meow!";
    }
}

class AnimalFactory
{
    public static Animal CreateAnimal(string animalType)
    {
        if (animalType == "dog")
            return new Dog();
        else if (animalType == "cat")
            return new Cat();
        else
            return null;
    }
}

class Program
{
    static void Main()
    {
        Animal dog = AnimalFactory.CreateAnimal("dog");
        if (dog != null) Console.WriteLine(dog.Speak()); // Woof!

        Animal cat = AnimalFactory.CreateAnimal("cat");
        if (cat != null) Console.WriteLine(cat.Speak()); // Meow!
    }
}

File: kotlin/factory_method.kt
abstract class Animal {
    abstract fun speak(): String
}

class Dog : Animal() {
    override fun speak(): String {
        return "Woof!"
    }
}

class Cat : Animal() {
    override fun speak(): String {
        return "Meow!"
    }
}

object AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

fun main() {
    val dog = AnimalFactory.createAnimal("dog")
    println(dog?.speak()) // Woof!

    val cat = AnimalFactory.createAnimal("cat")
    println(cat?.speak()) // Meow!
}
